File: frontend/src/types/advanced.ts
// Shapes mirror backend/app/models/advanced.py. When the backend schema
// changes, update this file in lockstep.

export type AdvancedMode =
  | 'aeo'
  | 'compare'
  | 'crawlTest'
  | 'authority'
  | 'citation'
  | 'visibility'
  | 'entity';

// ---- Requests ----

export interface CompareRequest {
  urls: string[];
}

export interface SingleUrlRequest {
  url: string;
}

export interface AiVisibilityRequest {
  url: string;
  custom_queries?: string[];
}

export interface EntityAuditRequest {
  entity_name: string;
  entity_type: 'brand' | 'product' | 'person';
}

export type AdvancedRequestBody<M extends AdvancedMode> = M extends 'compare'
  ? CompareRequest
  : M extends 'aeo'
  ? SingleUrlRequest
  : M extends 'visibility'
  ? AiVisibilityRequest
  : M extends 'entity'
  ? EntityAuditRequest
  : SingleUrlRequest;

// ---- Responses ----

export interface CompareUrlResult {
  url: string;
  domain: string;
  score: number;
  grade: string;
  categories: Record<string, { earned: number; max: number }>;
}

export interface CompareAdvantage {
  category: string;
  ranked: Array<{ domain: string; percent: number }>;
}

export interface CompareResponse {
  urls: string[];
  results: CompareUrlResult[];
  categories: string[];
  advantages: CompareAdvantage[];
  winner?: { domain: string; score: number; grade: string; lead: number } | null;
}

export interface CrawlTestResponse {
  url: string;
  domain: string;
  robots: {
    found: boolean;
    bots: Array<{ bot: string; ua: string; status: string }>;
    blocked: string[];
  };
  waf: {
    baseline_status: number | null;
    baseline_size: number;
    bots: Array<{
      bot: string;
      status_code: number | null;
      size: number;
      result: string;
      error: string | null;
    }>;
    blocked: string[];
  };
  common_crawl: {
    found: boolean;
    count: number;
    samples: string[];
    status: number | null;
    error: string | null;
  };
  total_issues: number;
}

export interface AuthorityAuditResponse {
  url: string;
  domain: string;
  brand: string;
  score: number;
  max_score: number;
  percent: number;
  grade: string;
  reviews: { platforms: string[]; has_schema: boolean };
  awards: {
    keywords: string[];
    has_schema: boolean;
    badges: string[];
    signal_count: number;
  };
  google: {
    indexed_pages: number | null;
    kg_schema_fields: number;
    wikipedia_or_wikidata: boolean;
    score: number;
  };
  mentions: { platforms: string[] };
}

export interface CitationCheckResponse {
  url: string;
  domain: string;
  brand: string;
  engine: string;
  total_queries: number;
  valid_queries: number;
  cited_queries: number;
  citation_rate: number;
  total_citations: number;
  grade: string;
  queries: Array<{
    query: string;
    cited: boolean;
    citations: string[];
    mentioned?: boolean;
    total_sources?: number;
    error: boolean;
  }>;
}

export interface AiVisibilityResponse {
  url: string;
  domain: string;
  brand: string;
  engines: string[];
  scores: {
    visibility: number;
    entity: number;
    competitor: number;
    stability: number;
    content_gap: number;
  };
  total_score: number;
  max_score: number;
  grade: string;
  grade_label: string;
  per_engine_rates: Record<string, number>;
  top_competitors: Array<{ domain: string; mentions: number }>;
  framings: Record<string, number>;
  content_gaps: string[];
  query_count: number;
  stability_runs: number;
}

export interface EntityAuditResponse {
  entity: string;
  entity_type: string;
  scores: Record<string, number>;
  total_score: number;
  max_score: number;
  percent: number;
  grade: string;
  knowledge_graph: {
    wikipedia: boolean;
    wikidata: boolean;
    wikidata_id: string | null;
    google_kg: boolean;
    platforms_found: string[];
  };
  platforms: { found: string[]; not_found: string[] };
  sentiment: string;
  best_framing: string;
  content_gaps: string[];
  recognition_rate: number;
  stability_runs: number;
}

export interface AeoVisibilityResponse {
  url: string;
  domain: string;
  score: number;
  grade: string;
  max_score: number;
  categories: Record<string, { earned: number; max: number }>;
  page_results: Array<{
    path: string;
    score: number;
    weakest_signal: string;
    weakest_detail: string;
  }>;
  priority_improvements: Array<{
    category: string;
    percent: number;
  }>;
}

export type AdvancedResponseOf<M extends AdvancedMode> = M extends 'compare'
  ? CompareResponse
  : M extends 'aeo'
  ? AeoVisibilityResponse
  : M extends 'crawlTest'
  ? CrawlTestResponse
  : M extends 'authority'
  ? AuthorityAuditResponse
  : M extends 'citation'
  ? CitationCheckResponse
  : M extends 'visibility'
  ? AiVisibilityResponse
  : M extends 'entity'
  ? EntityAuditResponse
  : never;

File: backend/geo/models/advanced.py
"""Pydantic request + response models for the 6 advanced GEO check modes.

Each mode gets its own request + response type. Responses are intentionally
flat so the frontend can render them with minimal transformation.
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any


# ---------------------------------------------------------------------------
# Requests
# ---------------------------------------------------------------------------

class CompareRequest(BaseModel):
    """--compare: 2+ URLs, returns side-by-side category breakdown."""
    urls: List[str] = Field(..., min_length=2, max_length=5)


class CrawlTestRequest(BaseModel):
    """--crawl-test: one URL, simulates AI bot access + Common Crawl lookup."""
    url: str


class AuthorityAuditRequest(BaseModel):
    """--authority-audit: one URL, off-page authority signals."""
    url: str


class CitationCheckRequest(BaseModel):
    """--citation-check: one URL, queries Perplexity for brand citations."""
    url: str


class AeoVisibilityRequest(BaseModel):
    """--aeo-visibility: one URL, AEO content optimization audit (free)."""
    url: str


class AiVisibilityRequest(BaseModel):
    """--ai-visibility: one URL, optional custom queries for the audit."""
    url: str
    custom_queries: Optional[List[str]] = None


class EntityAuditRequest(BaseModel):
    """--entity: entity name + type (brand|product|person)."""
    entity_name: str = Field(..., min_length=1)
    entity_type: str = Field(default="brand")


# ---------------------------------------------------------------------------
# Responses
# ---------------------------------------------------------------------------

class CompareUrlResult(BaseModel):
    url: str
    domain: str
    score: int
    grade: str
    categories: Dict[str, Any]


class CompareAdvantage(BaseModel):
    category: str
    ranked: List[Dict[str, Any]]


class CompareWinner(BaseModel):
    domain: str
    score: int
    grade: str
    lead: int


class CompareResponse(BaseModel):
    urls: List[str]
    results: List[CompareUrlResult]
    categories: List[str]
    advantages: List[CompareAdvantage]
    winner: Optional[CompareWinner] = None


class CrawlTestRobotEntry(BaseModel):
    bot: str
    ua: str
    status: str  # allowed | blocked


class CrawlTestWafEntry(BaseModel):
    bot: str
    status_code: Optional[int] = None
    size: int = 0
    result: str  # allowed | blocked | suspicious | error | unknown
    error: Optional[str] = None


class CrawlTestRobotsBlock(BaseModel):
    found: bool
    bots: List[CrawlTestRobotEntry]
    blocked: List[str]


class CrawlTestWafBlock(BaseModel):
    baseline_status: Optional[int] = None
    baseline_size: int = 0
    bots: List[CrawlTestWafEntry]
    blocked: List[str]


class CrawlTestCommonCrawl(BaseModel):
    found: bool = False
    count: int = 0
    samples: List[str] = Field(default_factory=list)
    status: Optional[int] = None
    error: Optional[str] = None


class CrawlTestResponse(BaseModel):
    url: str
    domain: str
    robots: CrawlTestRobotsBlock
    waf: CrawlTestWafBlock
    common_crawl: CrawlTestCommonCrawl
    total_issues: int


class AuthorityReviewsBlock(BaseModel):
    platforms: List[str]
    has_schema: bool


class AuthorityAwardsBlock(BaseModel):
    keywords: List[str]
    has_schema: bool
    badges: List[str]
    signal_count: int


class AuthorityGoogleBlock(BaseModel):
    indexed_pages: Optional[int] = None
    kg_schema_fields: int
    wikipedia_or_wikidata: bool
    score: int


class AuthorityMentionsBlock(BaseModel):
    platforms: List[str]


class AuthorityAuditResponse(BaseModel):
    url: str
    domain: str
    brand: str
    score: int
    max_score: int
    percent: float
    grade: str
    reviews: AuthorityReviewsBlock
    awards: AuthorityAwardsBlock
    google: AuthorityGoogleBlock
    mentions: AuthorityMentionsBlock


class CitationQueryResult(BaseModel):
    query: str
    cited: bool
    citations: List[str] = Field(default_factory=list)
    mentioned: bool = False
    total_sources: int = 0
    error: bool = False


class CitationCheckResponse(BaseModel):
    url: str
    domain: str
    brand: str
    engine: str
    total_queries: int
    valid_queries: int
    cited_queries: int
    citation_rate: float
    total_citations: int
    grade: str
    queries: List[CitationQueryResult]


class AiVisibilityScores(BaseModel):
    visibility: int
    entity: int
    competitor: int
    stability: int
    content_gap: int


class AiVisibilityCompetitor(BaseModel):
    domain: str
    mentions: int


class AiVisibilityResponse(BaseModel):
    url: str
    domain: str
    brand: str
    engines: List[str]
    scores: AiVisibilityScores
    total_score: int
    max_score: int
    grade: str
    grade_label: str
    per_engine_rates: Dict[str, float]
    top_competitors: List[AiVisibilityCompetitor]
    framings: Dict[str, int]
    content_gaps: List[str]
    query_count: int
    stability_runs: int


class EntityKnowledgeGraph(BaseModel):
    wikipedia: bool
    wikidata: bool
    wikidata_id: Optional[str] = None
    google_kg: bool
    platforms_found: List[str]


class EntityPlatforms(BaseModel):
    found: List[str]
    not_found: List[str]


class EntityAuditResponse(BaseModel):
    entity: str
    entity_type: str
    scores: Dict[str, int]
    total_score: int
    max_score: int
    percent: int
    grade: str
    knowledge_graph: EntityKnowledgeGraph
    platforms: EntityPlatforms
    sentiment: str
    best_framing: str
    content_gaps: List[str]
    recognition_rate: float
    stability_runs: int


class AeoCategoryScore(BaseModel):
    earned: float
    max: float


class AeoPageResult(BaseModel):
    path: str
    score: int
    weakest_signal: str
    weakest_detail: str


class AeoPriorityImprovement(BaseModel):
    category: str
    percent: int


class AeoVisibilityResponse(BaseModel):
    url: str
    domain: str
    score: int
    grade: str
    max_score: int
    categories: Dict[str, AeoCategoryScore]
    page_results: List[AeoPageResult]
    priority_improvements: List[AeoPriorityImprovement]


# ---------------------------------------------------------------------------
# Tier gating — mirrors frontend ADVANCED_MODES in Result.tsx. Keep in sync.
# ---------------------------------------------------------------------------

TIER_RANK: Dict[str, int] = {
    "free": 0,
    "pro": 1,
    "starter": 2,
    "growth": 3,
    "scale": 4,
}

# Each advanced mode's minimum paid tier. These match Result.tsx ADVANCED_MODES
# so frontend and backend agree on what's locked.
MODE_MIN_TIER: Dict[str, str] = {
    "aeo": "free",
    "compare": "pro",
    "crawlTest": "pro",
    "authority": "pro",
    "citation": "pro",
    "visibility": "pro",
    "entity": "pro",
}
